### PR TITLE
bpo-36097: Use only public C-API in the_xxsubinterpreters module (adding as necessary).

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
https://bugs.python.org/issue36097

### CODE REVIEW

File: Include/cpython/interpreteridobject.h [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#ifndef Py_CPYTHON_INTERPRETERIDOBJECT_H
+#  error "this header file must not be included directly"
+#endif
+
+/* Interpreter ID Object */
```
Comment (zooba) [inline Include/cpython/interpreteridobject.h:9]:
```c
#ifdef __cplusplus
extern "C" {
#endif
```
?

File: Include/cpython/pystate.h [hunk 2]
```diff
@@ -212,6 +213,63 @@ PyAPI_FUNC(PyThreadState *) PyThreadState_Next(PyThreadState *);
...
+    // likely a registered "crossinterpdatafunc", is responsible for
+    // ensuring it owns the reference (i.e. incref).
+    PyObject *obj;
+    // interp is the ID of the owning interpreter of the original
+    // object.  It corresponds to the active interpreter when
+    // _PyObject_GetCrossInterpreterData() was called.  This should only
+    // be set by the cross-interpreter machinery.
+    //
+    // We use the ID rather than the PyInterpreterState to avoid issues
+    // with deleted interpreters.
```
Comment (zooba) [inline Include/cpython/pystate.h]: Are IDs ever reused? (I think no, but worth repeating in context here.)

File: Include/cpython/pystate.h [hunk 3]
```diff
@@ -212,6 +213,63 @@ PyAPI_FUNC(PyThreadState *) PyThreadState_Next(PyThreadState *);
...
+PyAPI_FUNC(PyObject *) _PyCrossInterpreterData_NewObject(_PyCrossInterpreterData *);
+PyAPI_FUNC(void) _PyCrossInterpreterData_Release(_PyCrossInterpreterData *);
+
+PyAPI_FUNC(int) _PyObject_CheckCrossInterpreterData(PyObject *);
+
+/* cross-interpreter data registry */
+
+typedef int (*crossinterpdatafunc)(PyObject *, struct _xid *);
+
+PyAPI_FUNC(int) _PyCrossInterpreterData_Register_Class(PyTypeObject *, crossinterpdatafunc);
```
Comment (zooba) [inline Include/cpython/pystate.h]: I don't think we normally use extra underscores in C function names (`RegisterClass` vs `Register_Class`)

**Review by zooba — COMMENTED** [review]
A few comments as you work.

The title is a little deceptive, right? Most of this change is creating the public C API you need for this :)

File: Include/cpython/pystate.h (hunk 3 above)
Comment (ericsnowcurrently) [reply Include/cpython/pystate.h]: Whoa, I guess that one slipped through. :)

File: Include/cpython/pystate.h (hunk 2 above)
Comment (ericsnowcurrently) [reply Include/cpython/pystate.h]: They are never re-used.  There are 32 bits of them and we raise RuntimeError if you try to create more interpreters than that.

I've updated the comment.

File: Include/cpython/interpreteridobject.h (hunk 1 above)
Comment (ericsnowcurrently) [reply Include/cpython/interpreteridobject.h:9]: Hmm, I could swear Victor had left that out of the other header files in Include/cpython.  Maybe I'm starting to lose it. :)

fixed

**Comment by ericsnowcurrently** [comment]
@zooba, yeah, that title was unclear. :)  That's what happens when I get neck-deep in a space and why reviews are so important. :)

**Pull request merged**